madara-prover-rpc-server/src/services/starknet_prover.rs

```diff
@@ -0,0 +1,224 @@
+
+use madara_prover_common::models::{Proof, ProverConfig};
+use stone_prover::error::ProverError;
+
+use crate::cairo::{extract_execution_artifacts, ExecutionArtifacts, ExecutionError};
+use crate::services::common::{call_prover, format_prover_error, get_prover_parameters};
+use crate::services::starknet_prover::starknet_prover_proto::starknet_prover_server::StarknetProver;
+use crate::services::starknet_prover::starknet_prover_proto::{
+    StarknetExecutionRequest, StarknetProverResponse,
+};
+
+pub mod starknet_prover_proto {
+    tonic::include_proto!("starknet_prover");
+}
+
+const BOOTLOADER_PROGRAM: &[u8] =
+    include_bytes!("../../../test-cases/cases/bootloader/bootloader.json");
+
+// Copied from cairo_run.rs and adapted to support injecting the bootloader input.
+// TODO: check if modifying CairoRunConfig to specify custom variables is accepted upstream.
+pub fn cairo_run(
+    program: &Program,
+    cairo_run_config: &CairoRunConfig,
+    hint_executor: &mut dyn HintProcessor,
+    variables: HashMap<String, Box<dyn Any>>,
+) -> Result<(CairoRunner, VirtualMachine), CairoRunError> {
+    let secure_run = cairo_run_config
+        .secure_run
+        .unwrap_or(!cairo_run_config.proof_mode);
+
+    let mut cairo_runner = CairoRunner::new(
+        program,
+        cairo_run_config.layout,
+        cairo_run_config.proof_mode,
+    )?;
+    for (key, value) in variables {
+        cairo_runner.exec_scopes.insert_box(&key, value);
+    }
+
+    let mut vm = VirtualMachine::new(cairo_run_config.trace_enabled);
+    let end = cairo_runner.initialize(&mut vm)?;
+    // check step calculation
+
+    cairo_runner
+        .run_until_pc(end, &mut vm, hint_executor)
+        .map_err(|err| VmException::from_vm_error(&cairo_runner, &vm, err))?;
+    cairo_runner.end_run(
+        cairo_run_config.disable_trace_padding,
+        false,
+        &mut vm,
+        hint_executor,
+    )?;
+
+    vm.verify_auto_deductions()?;
+    cairo_runner.read_return_values(&mut vm)?;
+    if cairo_run_config.proof_mode {
+        cairo_runner.finalize_segments(&mut vm)?;
+    }
+    if secure_run {
+        verify_secure_runner(&cairo_runner, true, None, &mut vm)?;
+    }
+    cairo_runner.relocate(&mut vm, cairo_run_config.relocate_mem)?;
+
+    Ok((cairo_runner, vm))
+}
+
+#[derive(thiserror::Error, Debug)]
+enum BootloaderTaskError {
+    #[error("Failed to read program: {0}")]
+    Program(#[from] ProgramError),
+
+    #[error("Failed to read PIE: {0}")]
+    Pie(#[from] CairoPieError),
+}
+
+fn make_bootloader_tasks(
+    programs: &[Vec<u8>],
+    pies: &[Vec<u8>],
+) -> Result<Vec<TaskSpec>, BootloaderTaskError> {
+    let program_tasks = programs.iter().map(|program_bytes| {
+        let program = Program::from_bytes(program_bytes, Some("main"));
+        program
+            .map(|program| TaskSpec {
+                task: Task::Program(program),
+            })
+            .map_err(BootloaderTaskError::Program)
```

Comment:
This might lead to a frustrating debug experience since there's no additional context captured. Maybe `BootloaderTaksError::Program` (and `::Pie`) could take an index or something else to help identify which program/pie is problematic.

Reply:
A log statement could be a simpler way to help that

Reply:
Agreed, but I won't improve it right now. We're only trying to run a single PIE at this stage.